src/intptrcast.rs

```diff
@@ -260,14 +261,17 @@ impl<'mir, 'tcx> GlobalStateInner {
             Size::from_bytes(dl.overflowing_signed_offset(addr.bytes(), neg_base_addr).0),
         ))
     }
+}
 
-    /// Shifts `addr` to make it aligned with `align` by rounding `addr` to the smallest multiple
-    /// of `align` that is larger or equal to `addr`
-    fn align_addr(addr: u64, align: u64) -> u64 {
-        match addr % align {
-            0 => addr,
-            rem => addr.checked_add(align).unwrap() - rem,
-        }
+impl GlobalStateInner {
+    pub fn free_alloc_id(&mut self, dead_id: AllocId) {
+        // We can *not* remove this from `base_addr`, since `addr_from_alloc_id` is called on each
+        // attempt at a memory access to determine the allocation ID and offset -- and there can
+        // still be pointers with `dead_id` that one can attempt to use for a memory access.
```

Comment:
I don't understand why this is problematic for accesses. Even if we remove an allocation which still has pointers to it, the pointer is still dangling right? I expected the problem here to be converting pointers to freed allocations to integers.

Reply:
When such a dangling pointer gets used, the first thing that happens is that we call `ptr_get_alloc`, which is expected to return the `AllocId` and the offset inside the allocation. Once we remove an allocation from `base_addr` we can no longer determine the offset, so we have a problem.

Reply:
In what case would we need to know the offset inside the allocation and also be okay with using a pointer to a deallocated allocation?

Reply:
You are implicitly suggesting a rework of the `Machine` trait here. And it's not super trivial since in error messages we want to distinguish "ptr with no provenance" from "ptr with provenance of a dangling allocation", so just returning `None` when asked for "what allocid + offset is this provenance" would not work.

So this can probably be done but it's a bigger change than what this PR does. So far the API is designed around the idea that "get allocid + offset" and "determine whether alloc is live and offset inbounds" are two separate independent steps.

Reply:
> You are implicitly suggesting a rework of the Machine trait here.

I didn't mean to suggest that. I think this whole situation you're describing is part of why I was so confused, I don't know the interpreter internals well enough to reason from them as opposed to from the language. And to this specific question of mine above, I was half expecting you to produce some obscure Rust code in answer.

Can you add a short comment to indicate that the design of the interpreter requires us to be able to do "get alloc_id + offset" _before_ we do UB detection? Perhaps something like

```rust
        // We can *not* remove this from `base_addr`, since the interpreter design requires that we be able to retrieve
        // an AllocId + offset for any memory access *before* we check if the access is valid.
```

Reply:
r=me with any such tweak

Reply:
Ah okay. :) I incorporated that into the comment.